ambry-rest/src/test/java/com.github.ambry.rest/NettyResponseChannelTest.java

```diff
@@ -578,6 +514,55 @@ private void checkHeaders(HttpRequest request, HttpResponse response)
         HttpHeaders.getHeader(request, MockNettyMessageProcessor.CUSTOM_HEADER_NAME),
         HttpHeaders.getHeader(response, MockNettyMessageProcessor.CUSTOM_HEADER_NAME));
   }
+
+  // keepAliveTest() helpers.
+
+  /**
+   * Does the keep-alive test by setting the {@link HttpHeaders.Names#CONNECTION} to its two possible values and tests
+   * that the response has the appropriate value for the {@link HttpHeaders.Names#CONNECTION}.
+   * @param channel the {@link EmbeddedChannel} to send the request over.
+   * @param httpMethod the {@link HttpMethod} of the request.
+   * @param errorCode the {@link RestServiceErrorCode} to induce at {@link MockNettyMessageProcessor}. {@code null} if
+   *                  {@link TestingUri#OnResponseCompleteWithNonRestException} is desired.
+   * @param expectedResponseStatus the expected {@link HttpResponseStatus} from remote.
+   * @return the {@link EmbeddedChannel} to use once this function is complete. If the channel did not close, this
+   * function will return the {@code channel} instance that was passed, otherwise it returns a new channel.
+   */
+  private EmbeddedChannel doKeepAliveTest(EmbeddedChannel channel, HttpMethod httpMethod,
+      RestServiceErrorCode errorCode, HttpResponseStatus expectedResponseStatus) {
+    boolean keepAlive = true;
+    for (int i = 0; i < 2; i++) {
+      HttpHeaders httpHeaders = new DefaultHttpHeaders();
+      TestingUri uri = TestingUri.OnResponseCompleteWithNonRestException;
+      if (errorCode != null) {
+        uri = TestingUri.OnResponseCompleteWithRestException;
+        httpHeaders.set(MockNettyMessageProcessor.REST_SERVICE_ERROR_CODE_HEADER_NAME, errorCode);
+      }
+      if (!keepAlive) {
+        httpHeaders.set(HttpHeaders.Names.CONNECTION, HttpHeaders.Values.CLOSE);
+      }
+      channel.writeInbound(RestTestUtils.createRequest(httpMethod, uri.toString(), httpHeaders));
+      HttpResponse response = (HttpResponse) channel.readOutbound();
+      assertEquals("Unexpected response status", expectedResponseStatus, response.getStatus());
+      if (!(response instanceof FullHttpResponse)) {
+        // empty the channel
+        while (channel.readOutbound() != null) {
+        }
+      }
+      boolean shouldBeAlive = keepAlive &&
+          !httpMethod.equals(HttpMethod.POST) && !NettyResponseChannel.CLOSE_CONNECTION_ERROR_STATUSES
```

Comment:
[This](https://github.com/linkedin/ambry/pull/280#discussion_r62243598) discussion should contain all the details. Feel free to ask for more clarifications :)